### ISSUE TITLE
Enabling disabling tests

### ISSUE THREAD

**Issue by waltertamboer** [phpab/phpab#21]
Should it be possible to enable and disable tests? One might want to disable a test but keep the code.

Also see the comment of @pachico in https://github.com/phpab/phpab/issues/17#issuecomment-203298204

**Comment by pachico** [comment]
Wouldn't that be a policy of a ParticipationStrategyInterface implementation?
Something like:

``` php
class ParticipationStrategyDisabled implements ParticipationStrategyInterface {
    public function shouldParticipate(){
        return false;
    }
}
``
In cascade, whenever ParticipationStrategyInterface returns false, storage won't save new values if they don't exist and keep current ones as they are.
How do you see it?
```

**Comment by waltertamboer** [comment]
Does that mean that you should add a chained participation strategy and add the Disabled strategy to it by default?

**Comment by pachico** [comment]
Well, that's another approach but I'm also fine with it.

**Comment by waltertamboer** [comment]
Is that not what you mean? What do you mean than?

**Comment by pachico** [comment]
I mean that it's not require do be a chainable.
According to the examples we've seen, something like:

``` php

// ParticipationStrategy implementation
class ParticipationStrategyDisabled implements ParticipationStrategyInterface {
    public function shouldParticipate(){
        return false;
    }
}

// Test definition
$test = new \Phpab\Phpab\Test(
    'foo_test',
    new \Phpab\Phpab\ParticipationStrategy\ParticipationStrategyDisabled(),
    new \Phpab\Phpab\VariantChooser\RandomVariantChooser()
);

```

**Comment by waltertamboer** [comment]
But that means I need to switch out the participation strategy with a custom one if I want to disable my test. If I than want to enable the test again, I need to look in my VCS to see what the original strategy was. Wouldn't it be more easy to just have a boolean flag to indicate whether or not the test is enabled?

**Comment by pachico** [comment]
I see what you mean and it's a good point.
So, we could:
- have in implementation for ParticipationStrategyContainerInterface that would be ParticipationStrategyChaibable, that contains all other strategies (sort of container) where there could be a method Disable() (or other name) that you can set or not call, right?
- or have that ParticipationStrategyInterface had a method, as you say, setDisabled().

``` php
(new ParticipationStrategyFoo())->setDisabled();
```

is this what you mean?

**Comment by HellPat** [comment]
I think it shoul'd not be possible to disable Variants but only in thee Test itself.
At this point I am not sure about my Interface definition.

Maybe we should remove `TestInterface:::getParticipationStrategy` and `TestInterface::getVariantChooser` and add it to the definition of `EngineInterface::addTest` as Parameter. Maybe we should think about that.

**Comment by pachico** [comment]
VariantParticipationStrategy should be set per Test, I think, and not in the Engine itself (you could have different tests with different ones).
Doing it in the Test, however, seems ok. 
Any test, could be disabled independently from its strategy.

**Comment by waltertamboer** [comment]
What I mean is that I can have a full definition of a test but can simply disable it. This will be useful later on if we create integrations with frameworks. For ZF2 I would define all my tests in the config. In such a case it would be nice to simply disable a test:

``` php
<?php

// Pseudo code
return [
    'phpab' => [
        'storage' => 'cookie',
        'tests' => [
            'my-awesome-test' => [
                'enabled' => true,
                'participation_strategy' => 'Application\PhpAb\Participation\MyAwesomeStrategy',
                'options' => [
                    GExperiments::KEY => 'my-google-idenfier',
                ],
            ],
        ],
    ],
];
```

**Comment by pachico** [comment]
Yup, so at Test level and not ParticipationStrategy.

**Comment by waltertamboer** [comment]
Correct

**Comment by waltertamboer** [comment]
What do you think @psren ?

**Comment by HellPat** [comment]
Regarding to my last comment we should move `TestInterface:::getParticipationStrategy`and `TestInterface:::getVariantChooser` away from the test and instead "bloat" the addTest-Method on EngineInterface.

``` php
EngineInterface::addTest(
    TestInterface $test,
    ParticipationStrategyInterface $strategy,
    VariantChooserInterface $chooser
)
```

This way we can have the test in multiple engines (maybe for grouping) and different enabled states per engine. I don't know if it makes sense.

`DisabledStrategy` seems to be correct

**Comment by pachico** [comment]
I'm just trying to figure out for what reason someone would need two engines at all.

**Comment by HellPat** [comment]
@waltertamboer and me talked about Grouping of test.
Like "The User should only participate on one Test of Group A".

We already have kind of grouping. (in the engine).

So I'm not sure yet how we can implement it. It could be TestInterface or EngineInterface.
Nevertheless I'm seeing TestInterface getting too big and am more and more concerned that it was a bad descision to move strategy/chooser there.

We MUST have the possibility to defer these on a per test base. BUT the test should not know about it in a SOLID point of view.

**Comment by HellPat** [comment]
We could implement som kind of TestBag for internal usage in the runner. The Bag knows about Strategy and Chooser. We will get a better API i think :-D

**Comment by waltertamboer** [comment]
@psren Sounds like a good idea. Let's move it to a different issue so we don't have multiple discussions in one issue :)

About the enabling/disabling of tests. I still prefer a getEnabled method in TestInterface. Having a DisabledStrategy in my eyes makes it too complicated. It automatically means you need to setup a chained strategy or something in order to simply disable a test.

**Comment by HellPat** [comment]
I agree I renamed the StrategyInterface. I think with the new definition my own confusion is resolved :-D

**Comment by HellPat** [comment]
Everything in here. https://github.com/phpab/phpab/pull/29
Needs to be implemented though

**Comment by HellPat** [comment]
We could easily add a second parameter [here](https://github.com/phpab/phpab/blob/master/src/Engine/Engine.php#L79)  `callable $filter = null` and check the options of the test [here](https://github.com/phpab/phpab/blob/master/src/Engine/Engine.php#L84)
